packages/lit-html/src/lib/directives/repeat.ts

```diff
@@ -0,0 +1,445 @@
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {directive, NodePart, Part, Directive, noChange} from '../lit-html.js';
+import {
+  createAndInsertPart,
+  insertPartBefore,
+  removePart,
+  updatePart,
+} from '../parts.js';
+
+export type KeyFn<T> = (item: T, index: number) => unknown;
+export type ItemTemplate<T> = (item: T, index: number) => unknown;
+
+// Helper for generating a map of array item to its index over a subset
+// of an array (used to lazily generate `newKeyToIndexMap` and
+// `oldKeyToIndexMap`)
+const generateMap = (list: unknown[], start: number, end: number) => {
+  const map = new Map();
+  for (let i = start; i <= end; i++) {
+    map.set(list[i], i);
+  }
+  return map;
+};
+
+/**
+ * A directive that repeats a series of values (usually `TemplateResults`)
+ * generated from an iterable, and updates those items efficiently when the
+ * iterable changes based on user-provided `keys` associated with each item.
+ *
+ * Note that if a `keyFn` is provided, strict key-to-DOM mapping is maintained,
+ * meaning previous DOM for a given key is moved into the new position if
+ * needed, and DOM will never be reused with values for different keys (new DOM
+ * will always be created for new keys). This is generally the most efficient
+ * way to use `repeat` since it performs minimum unnecessary work for insertions
+ * and removals.
+ *
+ * IMPORTANT: If providing a `keyFn`, keys *must* be unique for all items in a
+ * given call to `repeat`. The behavior when two or more items have the same key
+ * is undefined.
+ *
+ * If no `keyFn` is provided, this directive will perform similar to mapping
+ * items to values, and DOM will be reused against potentially different items.
+ */
+
+class RepeatDirective extends Directive {
+  itemKeys?: unknown[];
+
+  constructor(part: Part) {
+    super();
+    if (!(part instanceof NodePart)) {
+      throw new Error('repeat can only be used in text bindings');
+    }
+  }
+
+  render<T>(
+    items: Iterable<T>,
+    keyFnOrTemplate: KeyFn<T> | ItemTemplate<T>,
+    template?: ItemTemplate<T>
+  ) {
+    let keyFn: KeyFn<T> | undefined;
+    if (template === undefined) {
+      template = keyFnOrTemplate;
+    } else if (keyFnOrTemplate !== undefined) {
+      keyFn = keyFnOrTemplate as KeyFn<T>;
+    }
+    this.itemKeys = [];
+    const values = [];
+    let index = 0;
+    for (const item of items) {
+      this.itemKeys[index] = keyFn ? keyFn(item, index) : index;
+      values[index] = template!(item, index);
+      index++;
+    }
+    return values;
+  }
+
+  update<T>(
```

Comment:
Is this the same implementation as 1.x?

Reply:
Basically, except for the call to `this.render()`